File: packages/shared/src/suiji/reviews.ts
import type { RecordKind, TaskStatus } from "./records";

export type ReviewScope =
  | { kind: "all" | "open" }
  | { kind: "record"; recordId: string };
export type ReviewInput = {
  question: string;
  scope: ReviewScope;
  history?: Array<{ role: "user" | "assistant"; text: string }>;
};
export type ReviewCitation = {
  recordId: string;
  version: number;
  kind: RecordKind;
  taskStatus: TaskStatus | null;
  createdAt: string;
  quote: string;
  attachment?: {
    id: string;
    fileName: string;
    mimeType: string;
    offset?: number;
  };
};
export type ReviewAnswer = {
  text: string;
  citations: ReviewCitation[];
  coverage: {
    mode: "agent-keyword";
    scope: ReviewScope;
    listedRecords: number;
    readRecords: number;
    attachmentReads: number;
    semanticIndex: false;
    externalLinks: false;
  };
};
export type SuijiReview = {
  id: string;
  status: "running" | "completed" | "failed" | "cancelled";
  createdAt: string;
  answer?: ReviewAnswer;
  error?: string;
};

File: packages/suiji-ios/Sources/SuijiIOS/Contracts/Reviews.swift
import Foundation

struct ReviewCapability: Codable, Sendable { let enabled: Bool; let provider: String }
struct ReviewScope: Codable, Hashable, Sendable {
  let kind: String
  let recordId: String?
  static let all = ReviewScope(kind: "all", recordId: nil)
  static let open = ReviewScope(kind: "open", recordId: nil)
  static func record(_ id: String) -> ReviewScope { ReviewScope(kind: "record", recordId: id) }
}
struct ReviewInput: Encodable, Sendable {
  struct Message: Encodable, Sendable { let role: String; let text: String }
  let question: String
  let scope: ReviewScope
  let history: [Message]
}
struct ReviewCitation: Codable, Identifiable, Sendable {
  struct File: Codable, Sendable { let id: String; let fileName: String; let mimeType: String; let offset: Int? }
  let recordId: String
  let version: Int
  let kind: RecordKind
  let taskStatus: TaskStatus?
  let createdAt: String
  let quote: String
  let attachment: File?
  var id: String { recordId + ":\(version):" + (attachment?.id ?? "") + quote }
}
struct ReviewAnswer: Codable, Sendable {
  struct Coverage: Codable, Sendable {
    let mode: String; let scope: ReviewScope; let listedRecords: Int; let readRecords: Int
    let attachmentReads: Int; let semanticIndex: Bool; let externalLinks: Bool
  }
  let text: String
  let citations: [ReviewCitation]
  let coverage: Coverage
}
struct SuijiReview: Codable, Identifiable, Sendable {
  let id: String
  let status: String
  let createdAt: String
  let answer: ReviewAnswer?
  let error: String?
}
